### PR TITLE
tests: Make descriptor tests deterministic

### PR DESCRIPTION
**Pull request by davereikher**
This is an improvement to a test, inspired by #14343 - removing non determinism from a test.

The test `descriptor_test` is non-deterministic, as it relies on the `MaybeUseHInsteadOfApostrophy` function which randomly either swaps all apostrophes with 'h' or doesn't at all in a descriptor. This fix makes both cases always run, if an apostrophe is found in a test descriptor.
This does not reduce test coverage but removes the non-determinism.

Additionally, the `MaybeUseHInsteadOfApostrophy` function removed the checksum if found at the end of a descriptor when the apostrophes are swapped by 'h's, since after being swapped the checksum is no longer correct. I instead added re-calculation of the checksum using the `DescriptorChecksum` function, which adds coverage for the case of a descriptors having 'h's instead of apostrophes and a checksum. This was previously lacking.
To achieve this I had to move `DescriptorChecksum` and `PolyMod` out of the anonymous namespace in descriptor.cpp to make `DescriptorChecksum` accessible in descriptor_tests.cpp.

All tests complete successfully (functional as well as unit tests).

### CODE REVIEW

**Comment by davereikher** [comment]
Woops, lint failed.. 
Is there an easy way to reproduce a Travis job locally? I'm a newb to Travis. I tried following https://stackoverflow.com/questions/21053657/how-to-run-travis-ci-locally/49019950 but I don't understand the instance field that is displayed in the Travis build logs (there is more than one instance there?)

**Comment by fanquake** [comment]
@davereikher The linter is just complaining about [trailing whitespace](https://travis-ci.org/bitcoin/bitcoin/jobs/569314353):
```bash
This diff appears to have added new lines with trailing whitespace.

The following changes were suspected:

diff --git a/src/test/descriptor_tests.cpp b/src/test/descriptor_tests.cpp

@@ -55,0 +63 @@ std::string MaybeUseHInsteadOfApostrophy(std::string ret)

+

@@ -61 +69,2 @@ const std::set<std::vector<uint32_t>> ONLY_EMPTY{{}};

+void DoCheck(const std::string& prv, const std::string& pub, int flags, const std::vector<std::vector<std::string>>& scripts, const std::set<std::vector<uint32_t>>& paths = ONLY_EMPTY,

^---- failure generated from test/lint/lint-whitespace.sh
```

You should be able to run the `test/lint/lint-whitespace.sh` script (as well as the other linters) locally.

**Comment by davereikher** [comment]
The `test/lint/lint-whitespace.sh` script finishes without errors locally. I would like to reproduce the Travis environment to avoid fixing and pushing multiple times. If not, I can just remove the trailing whitespace and try again.

File: src/script/descriptor.cpp [hunk 1]
```diff
@@ -137,6 +109,36 @@ std::string DescriptorChecksum(const Span<const char>& span)
     return ret;
 }
 
+
+namespace {
+
+////////////////////////////////////////////////////////////////////////////
+// Checksum                                                               //
```
Comment (achow101) [inline src/script/descriptor.cpp]: This comment shouldn't be moved.

File: src/test/descriptor_tests.cpp [hunk 2]
```diff
@@ -28,6 +31,8 @@ constexpr int HARDENED = 2; // Derivation needs access to private keys
 constexpr int UNSOLVABLE = 4; // This descriptor is not expected to be solvable
 constexpr int SIGNABLE = 8; // We can sign with this descriptor (this is not true when actual BIP32 derivation is used, as that's not integrated in our signing code)
 
+
```
Comment (achow101) [inline src/test/descriptor_tests.cpp]: Unnecessary whitespace change.

File: src/test/descriptor_tests.cpp [hunk 3]
```diff
@@ -40,32 +45,46 @@ bool EqualDescriptor(std::string a, std::string b)
...
 {
     FlatSigningProvider keys_priv, keys_pub;
     std::set<std::vector<uint32_t>> left_paths = paths;
 
+    std::unique_ptr<Descriptor> parse_priv;
+    std::unique_ptr<Descriptor> parse_pub;
     // Check that parsing succeeds.
-    auto parse_priv = Parse(MaybeUseHInsteadOfApostrophy(prv), keys_priv);
-    auto parse_pub = Parse(MaybeUseHInsteadOfApostrophy(pub), keys_pub);
+    if (replace_apostrophe_with_h)
```
Comment (achow101) [inline src/test/descriptor_tests.cpp]: I think it would make more sense to largely leave this function untouched and instead do the `'` to `h` change in your new `Check` function below. The only change here would be to remove `MaybeUseHInsteadOfApostrophy`

**Review by achow101 — COMMENTED** [review]
Concept ACK. I've left a few comments.

I think it makes sense to test the parsing of both the `'` and `h` key origin forms for all descriptors.

File: src/script/descriptor.cpp (hunk 1 above)
Comment (davereikher) [reply src/script/descriptor.cpp]: I'm not sure I understand. It's not quite clear in the diff, but the `// Checksum` comment was not moved. It remains after `namespace {`. The only comment that was moved is the one above `PolyMod`, describing the operation of that function, along with the function.

File: src/test/descriptor_tests.cpp (hunk 3 above)
Comment (davereikher) [reply src/test/descriptor_tests.cpp]: I agree and I did it that way initially but then, for example, this check failed:
```
std::string pub1 = parse_priv->ToString();
BOOST_CHECK(EqualDescriptor(pub, pub1));
```
since Descriptor::ToString returns a descriptor formatted with `'`, while pub has `h`, if I do it the way you suggest. I had to work around this by passing pub and prv without modification and introducing a boolean argument of whether to swap `'` with `h` only when passing them to the `Parse` function.

File: src/test/descriptor_tests.cpp (hunk 2 above)
Comment (davereikher) [reply src/test/descriptor_tests.cpp]: Thanks, fixed in commit d3e2d08a3be92cecb50c89c69871e61e0618e10b.

File: src/script/descriptor.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/script/descriptor.cpp]: In other words: The function does not need to move, only the `namespace {`

Comment (davereikher) [reply src/script/descriptor.cpp]: Thanks, fixed in commit 6819fbd741891081b2679e4b763bb88dc997ffe7. Is there a way to re-run AppVeyor? Looks like it timed out.

Comment (MarcoFalke) [reply src/script/descriptor.cpp]: You can re-run the ci by squashing your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

Comment (davereikher) [reply src/script/descriptor.cpp]: Ok, done in commit 0fc46f5b9262f016c42d2abe68d5125833f55416.

File: src/test/descriptor_tests.cpp [hunk 4]
```diff
@@ -40,32 +43,46 @@ bool EqualDescriptor(std::string a, std::string b)
...
+    {
+        parse_priv = Parse(prv, keys_priv);
+        parse_pub = Parse(pub, keys_pub);
+    }
+    else
+    {
+        parse_priv = Parse(UseHInsteadOfApostrophe(prv), keys_priv);
+        parse_pub = Parse(UseHInsteadOfApostrophe(pub), keys_pub);
```
Comment (achow101) [inline src/test/descriptor_tests.cpp]: This is actually backwards. When `replace_apostrophe_with_h` is false, we're swapping. I think it's supposed to be the other way around.

File: src/test/descriptor_tests.cpp [hunk 5]
```diff
@@ -40,32 +43,46 @@ bool EqualDescriptor(std::string a, std::string b)
...
-            if (it != std::string::npos) {
-                ret[it] = 'h';
-                if (ret.size() > 9 && ret[ret.size() - 9] == '#') ret = ret.substr(0, ret.size() - 9); // Changing apostrophe to h breaks the checksum
-            } else {
-                break;
+    while (true) {
+        auto it = ret.find("'");
+        if (it != std::string::npos) {
+            ret[it] = 'h';
+            if (ret.size() > 9 && ret[ret.size() - 9] == '#') {
```
Comment (meshcollider) [inline src/test/descriptor_tests.cpp]: This will recalculate the checksum after every replacement, it would be better to just recalculate it at the end

**Review by meshcollider — COMMENTED** [review]
Concept ACK

File: src/test/descriptor_tests.cpp (hunk 5 above)
Comment (davereikher) [reply src/test/descriptor_tests.cpp]: Thanks! Fixed in 9f9afcf1355e48d2b39ad9119415d658f5770ec5.

File: src/test/descriptor_tests.cpp (hunk 4 above)
Comment (davereikher) [reply src/test/descriptor_tests.cpp]: Thanks, you're absolutely right. Fixed in a8f49b7d53d33aeba8bd0965c21878795b36f99b.

**Review by promag — COMMENTED** [review]
Concept ACK.

`PolyMod` linkage changed, make it static?

File: src/test/descriptor_tests.cpp [hunk 6]
```diff
@@ -40,32 +43,49 @@ bool EqualDescriptor(std::string a, std::string b)
     return a == b;
 }
 
-std::string MaybeUseHInsteadOfApostrophy(std::string ret)
+std::string UseHInsteadOfApostrophe(std::string ret)
```
[the remark this answers is not in the document]
Comment (davereikher) [reply src/test/descriptor_tests.cpp]: It's in an anonymous namespace, or am I missing something?

File: src/test/descriptor_tests.cpp [hunk 7]
```diff
@@ -11,6 +11,9 @@
 #include <script/descriptor.h>
 #include <util/strencodings.h>
 
+
+std::string DescriptorChecksum(const Span<const char>& span);
```
[the remark this answers is not in the document]
Comment (davereikher) [reply src/test/descriptor_tests.cpp]: Done in commit 07c079a2fb49dca58a1d7e5e185a1cbc430381df.

File: src/test/descriptor_tests.cpp [hunk 8]
```diff
@@ -40,32 +43,49 @@ bool EqualDescriptor(std::string a, std::string b)
...
-            if (it != std::string::npos) {
-                ret[it] = 'h';
-                if (ret.size() > 9 && ret[ret.size() - 9] == '#') ret = ret.substr(0, ret.size() - 9); // Changing apostrophe to h breaks the checksum
-            } else {
-                break;
-            }
+    bool ret_changed = false;
+    while (true) {
+        auto it = ret.find("'");
+        if (it != std::string::npos) {
```
[the remark this answers is not in the document]
Comment (davereikher) [reply src/test/descriptor_tests.cpp]: Done in commit 07c079a2fb49dca58a1d7e5e185a1cbc430381df.

File: src/test/descriptor_tests.cpp [hunk 9]
```diff
@@ -40,32 +43,49 @@ bool EqualDescriptor(std::string a, std::string b)
...
 {
     FlatSigningProvider keys_priv, keys_pub;
     std::set<std::vector<uint32_t>> left_paths = paths;
 
+    std::unique_ptr<Descriptor> parse_priv;
+    std::unique_ptr<Descriptor> parse_pub;
     // Check that parsing succeeds.
-    auto parse_priv = Parse(MaybeUseHInsteadOfApostrophy(prv), keys_priv);
-    auto parse_pub = Parse(MaybeUseHInsteadOfApostrophy(pub), keys_pub);
+    if (replace_apostrophe_with_h)
```
[the remark this answers is not in the document]
Comment (davereikher) [reply src/test/descriptor_tests.cpp]: Done in commit 07c079a2fb49dca58a1d7e5e185a1cbc430381df.

File: src/test/descriptor_tests.cpp [hunk 10]
```diff
@@ -40,32 +43,49 @@ bool EqualDescriptor(std::string a, std::string b)
...
+    {
+        parse_priv = Parse(UseHInsteadOfApostrophe(prv), keys_priv);
+        parse_pub = Parse(UseHInsteadOfApostrophe(pub), keys_pub);
+    }
```
[the remark this answers is not in the document]
Comment (davereikher) [reply src/test/descriptor_tests.cpp]: Done in commit 07c079a2fb49dca58a1d7e5e185a1cbc430381df.

File: src/test/descriptor_tests.cpp [hunk 11]
```diff
@@ -164,6 +184,16 @@ void Check(const std::string& prv, const std::string& pub, int flags, const std:
     BOOST_CHECK_MESSAGE(left_paths.empty(), "Not all expected key paths found: " + prv);
 }
 
+void Check(const std::string& prv, const std::string& pub, int flags, const std::vector<std::vector<std::string>>& scripts, const std::set<std::vector<uint32_t>>& paths = ONLY_EMPTY)
+{
+    DoCheck(prv, pub, flags, scripts, paths); // Check original (with apostrophes)
+
+    if ((prv.find("'") != std::string::npos) || (pub.find("'") != std::string::npos))
```
[the remark this answers is not in the document]
Comment (davereikher) [reply src/test/descriptor_tests.cpp]: Done in commit 07c079a2fb49dca58a1d7e5e185a1cbc430381df. Also, applied the same change (`find("'")`->`find('\'')`) here:https://github.com/bitcoin/bitcoin/blob/07c079a2fb49dca58a1d7e5e185a1cbc430381df/src/test/descriptor_tests.cpp#L47

File: src/test/descriptor_tests.cpp [hunk 12]
```diff
@@ -164,6 +184,16 @@ void Check(const std::string& prv, const std::string& pub, int flags, const std:
...
+    {
+        DoCheck(prv, pub, flags, scripts, paths, true); // Check with apostrophes replaced by 'h'
```
[the remark this answers is not in the document]
Comment (davereikher) [reply src/test/descriptor_tests.cpp]: Done in commit 07c079a.

**Comment by davereikher** [comment]
> Concept ACK.
> 
> `PolyMod` linkage changed, make it static?

Done in commit 07c079a.

**Comment by davereikher** [comment]
Squashed commits.

File: src/test/descriptor_tests.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (promag) [reply src/test/descriptor_tests.cpp]: True, sorry.

[the remark this answers is not in the document]
Comment (promag) [reply src/test/descriptor_tests.cpp]: But make it `std::string UseHInsteadOfApostrophe(const std::string& ret)`. Also could avoid `ret` argument, sounds like it's an output/return argument.

[the remark this answers is not in the document]
Comment (davereikher) [reply src/test/descriptor_tests.cpp]: I believe that this line would fail should ret be made a const reference:
https://github.com/bitcoin/bitcoin/blob/0b90954416767f6213094f1f282e708c3e8a710b/src/test/descriptor_tests.cpp#L50
I guess I could change the signature to what you suggest and then assign `ret` to a local variable. Then there would still be a single copy operation so it's not less efficient or anything, but that would introduce an extra line of code. Do you think I should do it?
I agree that `ret` is not an indicative name. I think `desc` is better?

**Comment by meshcollider** [comment]
I believe this will be a lot simpler change now if you use `GetDescriptorChecksum` from #15986

Concept ACK anyhow

**Comment by davereikher** [comment]
> I believe this will be a lot simpler change now if you use `GetDescriptorChecksum` from #15986
> 
> Concept ACK anyhow

Great! Now that `GetDescriptorChecksum` is available I moved `DescriptorChecksum` and `PolyMod` back into the anonymous namespace and instead used `GetDescriptorChecksum` to calculate the checksum. This also introduces an indirect test to `GetDescriptorChecksum` and the underlying functions.
My branch seems to have failed an unrelated test on Travis while passing all tests (unit and functional, including the test that fails on Travis) locally. Is sporadic failure a known issue perhaps? Is there a way to re-run Travis?

**Comment by fanquake** [comment]
@davereikher Travis tests can occasionally fail sporadically. I've rebooted the [two failing instances](https://travis-ci.org/bitcoin/bitcoin/builds/573642446) for you.

File: src/test/descriptor_tests.cpp [hunk 13]
```diff
@@ -42,33 +42,44 @@ bool EqualDescriptor(std::string a, std::string b)
...
+        auto it = ret.find('\'');
+        if (it == std::string::npos) break;
+        ret[it] = 'h';
+    }
+
+    // GetDescriptorChecksum returns "" if the checksum exists but is bad.
+    // Switching apostrophes with 'h' breaks the checksum if it exists - recalculate it and replace the broken one.
+    if (GetDescriptorChecksum(ret) == "") {
+      ret = ret.substr(0, desc.size() - 9);
+      ret += std::string("#") + GetDescriptorChecksum(ret);
```
Comment (meshcollider) [inline src/test/descriptor_tests.cpp]: nit: indentation

File: src/test/descriptor_tests.cpp [hunk 14]
```diff
@@ -42,33 +42,44 @@ bool EqualDescriptor(std::string a, std::string b)
...
     FlatSigningProvider keys_priv, keys_pub;
     std::set<std::vector<uint32_t>> left_paths = paths;
     std::string error;
 
+    std::unique_ptr<Descriptor> parse_priv;
+    std::unique_ptr<Descriptor> parse_pub;
     // Check that parsing succeeds.
-    auto parse_priv = Parse(MaybeUseHInsteadOfApostrophy(prv), keys_priv, error);
-    auto parse_pub = Parse(MaybeUseHInsteadOfApostrophy(pub), keys_pub, error);
+    if (replace_apostrophe_with_h) {
```
Comment (meshcollider) [inline src/test/descriptor_tests.cpp]: There is actually a test case which used to be covered which is no longer ever covered - the use of `h` in the pub and `'` in the prv and vice versa (because there was a 50% chance of each being changed independently)

Comment (davereikher) [reply src/test/descriptor_tests.cpp]: Thanks for your review. You are right, I completely missed this. I added code to cover all four test cases (no replacement, prv only, pub only and prv+pub) in b9ee63c71b751fd67da777774ea8c0b27e7db2f8.

**Comment by davereikher** [comment]
> @davereikher Travis tests can occasionally fail sporadically. I've rebooted the [two failing instances](https://travis-ci.org/bitcoin/bitcoin/builds/573642446) for you.

@fanquake Travis failed again, could you please reboot the failing instance [here](https://travis-ci.org/bitcoin/bitcoin/builds/574691201)?

**Review by meshcollider — COMMENTED** [review]
utACK

**Review by achow101 — APPROVED** [review]
Code Review ACK b9ee63c71b751fd67da777774ea8c0b27e7db2f8

**Comment by fanquake** [comment]
@davereikher Thanks for sticking with this, nice first time contribution.

**Comment by davereikher** [comment]
Thanks a lot! I learned so much from just this small contribution and your reviews. Looking forward for more! :)

**Comment by practicalswift** [comment]
@davereikher Very nice to have you onboard and thanks for this first contribution! Hope to see more contributions from you.

If you want to take on non-determinism in tests you might want to take a look at `contrib/devtools/test_deterministic_coverage.sh`.

The following unit tests are known to have non-deterministic line coverage:

https://github.com/bitcoin/bitcoin/blob/3ca514ddb77253042877d1a72dfd3021c3de2812/contrib/devtools/test_deterministic_coverage.sh#L16-L36

Getting them deterministic in terms of line coverage would be great! More specifically that would allow us to reason about how a proposed change alters unit tests line coverage. Currently line coverage is a stochastic process :-)

**Comment by davereikher** [comment]
> @davereikher Very nice to have you onboard and thanks for this first contribution! Hope to see more contributions from you.
> 
> If you want to take on non-determinism in tests you might want to take a look at `contrib/devtools/test_deterministic_coverage.sh`.
> 
> The following unit tests are known to have non-deterministic line coverage:
> 
> https://github.com/bitcoin/bitcoin/blob/3ca514ddb77253042877d1a72dfd3021c3de2812/contrib/devtools/test_deterministic_coverage.sh#L16-L36
> 
> Getting them deterministic in terms of line coverage would be great! More specifically that would allow us to reason about how a proposed change alters unit tests line coverage. Currently line coverage is a stochastic process :-)

Thanks! Sounds great, I'll look into the non-determinism of those tests and I guess I will start with the low hanging fruit first :)
